### PR TITLE
[feature]: Automatically add test reports folder to gitignore

### PR DESCRIPTION
**Pull request by Akash-Singh04**
## Related Issue
  - Create a gitignore file in the keploy folder that adds the test report folder and prevents it from accidentally being commited.

Closes: #1775 

#### Describe the changes you've made
  - Create a gitignore file in the keploy folder that adds the test report folder and prevents it from accidentally being commited.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added
NIL
Please describe the tests(if any). Provide instructions how its affecting the coverage.
NIL
#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.
NIL
## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

image

### CODE REVIEW

**Comment by Akash-Singh04** [comment]
@slayerjain Kindly review this and let me know if this implementation works!

**Review by PranshuSrivastava — CHANGES_REQUESTED** [review]

File: pkg/service/record/record.go [hunk 1]
```diff
@@ -178,6 +180,15 @@ func (r *Recorder) Start(ctx context.Context) error {
 				r.telemetry.RecordedTestCaseMock(mock.GetKind())
 			}
 		}
+		//create  a gitignore file to ignore the test reports folder
+		gitIgnorePath := filepath.Join(r.config.Path, ".gitignore")
+		gitIgnoreData := []byte("./reports/\n")
+		err = os.WriteFile(gitIgnorePath, gitIgnoreData, 0777)
```
[the remark this answers is not in the document]
Comment (Akash-Singh04) [reply pkg/service/record/record.go:186]: Refactored to a util function and added it to record and test command. We are calling the util function only if there is no error in the previous insertmock call

File: pkg/util.go [hunk 2]
```diff
@@ -77,6 +77,30 @@ func IsTime(stringDate string) bool {
 	return false
 }
 
+func CreateGitIgnore(logger *zap.Logger, filePath string) error {
+	gitIgnoreData := []byte("./reports/\n")
```
Comment (PranshuSrivastava) [inline pkg/util.go:81]: The path of the reports directory is incorrect

File: pkg/util.go [hunk 3]
```diff
@@ -77,6 +77,30 @@ func IsTime(stringDate string) bool {
...
+			utils.LogError(logger, err, "failed to close gitignore file")
+		}
+	}()
+	//writing the data to the file
+	_, err = gitIgnoreFIle.Write(gitIgnoreData)
+	if err != nil {
+		utils.LogError(logger, err, "failed to write data to gitignore file")
+		return err
+	}
+	logger.Info("Gitignore file generated successfully")
```
Comment (PranshuSrivastava) [inline pkg/util.go:100]: Instead of this message, tell the user that the reports folder has been added to a gitignore file

File: pkg/util.go [hunk 4]
```diff
@@ -77,6 +77,30 @@ func IsTime(stringDate string) bool {
 	return false
 }
 
+func CreateGitIgnore(logger *zap.Logger, filePath string) error {
+	gitIgnoreData := []byte("./reports/\n")
+	filePath = filePath + "/.gitignore"
+	gitIgnoreFIle, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY, 0777)
```
Comment (PranshuSrivastava) [inline pkg/util.go:83]: what is the gitignore file exists already?

File: pkg/service/record/record.go [hunk 5]
```diff
@@ -178,6 +178,10 @@ func (r *Recorder) Start(ctx context.Context) error {
 				r.telemetry.RecordedTestCaseMock(mock.GetKind())
 			}
 		}
+		err := pkg.CreateGitIgnore(r.logger, r.config.Path)
```
Comment (PranshuSrivastava) [inline pkg/service/record/record.go:181]: Why do you need to create the gitignore file during the record mode? We only need to create it when reports are created.

File: pkg/util.go [hunk 6]
```diff
@@ -77,6 +77,30 @@ func IsTime(stringDate string) bool {
 	return false
 }
 
+func CreateGitIgnore(logger *zap.Logger, filePath string) error {
```
Comment (PranshuSrivastava) [inline pkg/util.go:80]: I think this is a generic function so should be added to global utils instead of pkg utils.

**Review by PranshuSrivastava — CHANGES_REQUESTED** [review]

File: pkg/util.go (hunk 2 above)
Comment (Akash-Singh04) [reply pkg/util.go:81]: tree -a
.
├── .gitignore
├── reports
│   ├── test-run-0
│   │   └── test-set-0-report.yaml
│   └── test-run-1
│       └── test-set-0-report.yaml
└── test-set-0
    ├── mocks.yaml
    └── tests
        ├── test-1.yaml
        ├── test-2.yaml
        └── test-3.yaml

5 directories, 7 files
If this is the folder structure of the keploy folder, could you please explain how is the path to the reports folder incorrect?

File: pkg/util.go (hunk 6 above)
Comment (Akash-Singh04) [reply pkg/util.go:80]: Added check for when gitignore file already exists

File: pkg/service/record/record.go (hunk 5 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go:196]: Apologies for the confusion with your previous comment 
> Why is this only in the record mode? What if the user starts the test mode directly by importing the testcases from somewhere?Also we need to check if the testreport was generated successfully, and only then add it to the gitignore file

Removed it from the record command

File: pkg/util.go (hunk 6 above)
Comment (Akash-Singh04) [reply pkg/util.go:80]: Refactored to global utils

**Comment by Akash-Singh04** [comment]
@PranshuSrivastava Your requested changes have been made

**Comment by gouravkrosx** [comment]
@Akash-Singh04 Please make a fresh PR for the same issue and close this PR

**Pull request closed without merging**